client/src/main/java/io/pravega/schemaregistry/client/SchemaRegistryClient.java

```diff
@@ -0,0 +1,364 @@
+import io.pravega.schemaregistry.contract.data.SchemaValidationRules;
+import io.pravega.schemaregistry.contract.data.SchemaWithVersion;
+import io.pravega.schemaregistry.contract.data.VersionInfo;
+
+import javax.annotation.Nullable;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import static io.pravega.schemaregistry.client.exceptions.RegistryExceptions.*;
+
+/**
+ * Defines a registry client for interacting with schema registry service. 
+ * The implementation of this interface should provide atomicity and read-after-write-consistency guarantees for all the methods.
+ */
+@Beta
+public interface SchemaRegistryClient {
+    /**
+     * Adds a new group. A group refers to the name under which the schemas are registered. A group is identified by a 
+     * unique name and has an associated set of group metadata {@link GroupProperties} and a list of codec types and a 
+     * versioned history of schemas that were registered under the group. 
+     * Add group is idempotent. If the group by the same name already exists the api will return false. 
+     * 
+     * @param groupId Id for the group that uniquely identifies the group. 
+     * @param groupProperties groupProperties Group properties for the group. These include serialization format, validation rules, 
+     *                        and flag to declare whether multiple schemas representing distinct object types can be 
+     *                        registered with the group. Type identify objects of same type. Schema compatibility checks 
+     *                        are always performed for schemas that share same {@link SchemaInfo#type}.
+     *                        Additionally, a user defined map of properties can be supplied.
+     * @return True indicates if the group was added successfully, false if it exists. 
+     * @throws BadArgumentException if the group properties is rejected by service.
+     * @throws UnauthorizedException if the user is unauthorized.
+     */
+    boolean addGroup(String groupId, GroupProperties groupProperties) throws BadArgumentException, UnauthorizedException;
+    
+    /**
+     * Removes a group identified by the groupId. This will remove all the codec types and schemas registered under the group.
+     * Remove group is idempotent. 
+     * 
+     * @param groupId Id for the group that uniquely identifies the group. 
+     * @throws UnauthorizedException if the user is unauthorized.
+     */
+    void removeGroup(String groupId) throws UnauthorizedException;
+
+    /**
+     * List all groups that the user is authorized on. This returns an iterator where each element is a pair of group 
+     * name and group properties. 
+     * This iterator can be used to iterate over each element until all elements are exhausted. 
+     * The implementation should guarantee that all groups added before and until the iterator returns 
+     * {@link Iterator#hasNext()} = true can be iterated over. 
```

Comment:
The idea is as follows:
request list of items from the service with a limit. 
The service returns some items and a continuation token. 
After exhausting the returned list of items, the caller calls into the service with the continuation token and if it gets a null/empty response, that is the exit condition for iteration. 

So if items are added "after" the position identified by continuation token, the subsequent request to get items from continuation token will have these new items returned. 

So the promise as described above would hold true.

Reply:
So suppose there are 500 items to be returned. `listGroups` is called. Then it is handed to a `for` loop to process. It is immediately going to call `hasNext` which is going to return `true` because there are more items. So after this point if an item is added is will not be seen by the iterator because it already returned `true`. 

That sounds wrong to me. Do you really not mean `false`?

Reply:
You make a call and get an iterator back. 
There are 500 items returned by the service along with a continuation token. 
The iterator is pointing to HEAD element. 
When you do iterator.hasNext, since HEAD is set the iterator will return true. 
You will keep getting true until you hit the 500th element. 
At this point, when you do hasNext, the iterator actually calls into the service with continuation token. If there were 10 more elements added, service will return those 10 new elements and a new continuation token. 
the iterator will return true 10 more times after that the call is again made to the service. 
this time service returns an empty set of items. 
iterator.hasNext returns false at this point. And iteration ends.

Reply:
Re-read the javadoc. It is not consistent with this explanation.